### ISSUE TITLE
net/http/httptest: NewRequest panics whereas net/http NewRequest does not

### ISSUE THREAD

**Issue by arthurgustin** [golang/go#26083]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?

`go version go1.10.2 linux/amd64`

### Does this issue reproduce with the latest release?

Yes on the go playground

### What operating system and processor architecture are you using (`go env`)?

```
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/agustin/.cache/go-build"
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/agustin/gocode"
GORACE=""
GOROOT="/usr/local/go"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build158217544=/tmp/go-build -gno-record-gcc-switches"
```

### What did you do?

In unit tests, I use httptest to simulate user requests.

### What did you expect to see?

The request properly created

### What did you see instead?

```
Test Panicked
                invalid NewRequest arguments; malformed HTTP version "((0 0, 0 10, 10 10, 10 0, 0 0),(3 3, 3 7, 7 7, 7 3, 3 3))&relation=contains HTTP/1.0"
                /usr/local/go/src/runtime/panic.go:502

                Full Stack Trace
                	/usr/local/go/src/runtime/panic.go:502 +0x229
                net/http/httptest.NewRequest(0x94b42a, 0x3, 0xc42002c4d0, 0x6a, 0x7f81c8858bd0, 0xc4209a14a0, 0x8e6800)
                	/usr/local/go/src/net/http/httptest/httptest.go:47 +0x690

```

### Playground link

https://play.golang.org/p/SB4a-aRS6xr

**Comment by odeke-em** [comment]
Thank you for the report @arthurgustin and thank you for the triage @agnivade!

@agnivade as we are quite late in the Go1.11 cycle, perhaps let's punt this off to Go1.12 or even make it `Unreleased`? This code has been like this for the past 2 years since March 2016 for Go1.7 as per https://github.com/golang/go/commit/694eadcce755e9d8b02a9fed8b79448cc8bbcd5c aka CL [21016](https://go-review.googlesource.com/c/go/+/21016)

Paging @bradfitz too as in https://github.com/golang/go/issues/14199 this was intended for outbound requests and using http.ReadRequest.

**Comment by odeke-em** [comment]
*inbound requests

**Comment by bradfitz** [comment]
This is working as intended. httptest.NewRequest is for incoming requests. And it's for tests, so panicking is fine, and the target represents the on-wire representation, which doesn't permit spaces. http.NewRequest is for outgoing requests, and represents a URL, not its wire representation.

**Comment by arthurgustin** [comment]
I can't test right now, but I tried to URL encode spaces and I got the same
error. Can you explain a bit more why this is not a bug ?

Le mer. 27 juin 2018 20:50, Brad Fitzpatrick <notifications@github.com> a
écrit :

> Closed #26083 <https://github.com/golang/go/issues/26083>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on
> <https://github.com/golang/go/issues/26083#event-1704665362>
>

GitHub <https://github.com/golang/go/issues/26083#event-1704665362>, or mute
